### PR TITLE
refactor: enable waiting for manual snapshot

### LINKED ISSUE #8686: AsyncSnapshotingTest.shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed
**Summary**

> Try to answer the following as best as possible

- How often does the test fail? [At least once](https://ci.zeebe.camunda.cloud/blue/organizations/jenkins/camunda-cloud%2Fzeebe/detail/release-1.4.0-alpha1/1/tests)
- Does it block your work? No
- Do we suspect that it is a real failure? I don't think so

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>

```
org.awaitility.core.ConditionTimeoutException: 
Assertion condition defined as a io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest 
Expecting Optional to contain a value but it was empty within 10 seconds.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:165)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:895)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:679)
	at io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest.shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed(AsyncSnapshotingTest.java:175)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.lambda$execute$1(JUnitPlatformProvider.java:199)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:193)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:120)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.lang.AssertionError: 
Expecting Optional to contain a value but it was empty.
	at io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest.lambda$shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed$5(AsyncSnapshotingTest.java:175)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:222)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:209)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>

```
Standard Output
12:55:36.922 [] [main] DEBUG io.camunda.zeebe.logstreams.snapshot - Opened database from '/tmp/junit5834930714341401105/runtime'.
12:55:37.132 [Broker-1-SnapshotStore-1] [-zb-fs-workers-1] DEBUG io.camunda.zeebe.logstreams.snapshot - Taking temporary snapshot into /tmp/junit5834930714341401105/pending/125-1-25-9223372036854775807.
12:55:37.234 [Broker-1-SnapshotStore-1] [-zb-fs-workers-0] DEBUG io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot - Aborting transient snapshot FileBasedTransientSnapshot{directory=/tmp/junit5834930714341401105/pending/125-1-25-9223372036854775807, checksum=2956754977, metadata=FileBasedSnapshotMetadata{index=125, term=1, processedPosition=25, exporterPosition=9223372036854775807}}
12:55:37.157 [Broker-0-SnapshotDirector-1] [-zb-fs-workers-1] ERROR io.camunda.zeebe.logstreams.snapshot - Unexpected error in resolving last written position.
java.lang.RuntimeException: getLastWrittenPositionAsync fails
	at io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest.shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed(AsyncSnapshotingTest.java:160) ~[test-classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.RunRules.evaluate(RunRules.java:20) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42) ~[junit-vintage-engine-5.8.2.jar:5.8.2]
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80) ~[junit-vintage-engine-5.8.2.jar:5.8.2]
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72) ~[junit-vintage-engine-5.8.2.jar:5.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.lambda$execute$1(JUnitPlatformProvider.java:199) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at java.util.Iterator.forEachRemaining(Iterator.java:133) ~[?:?]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:193) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:120) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
12:55:47.355 [] [main] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-fs-workers'
12:55:47.355 [] [main] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-actors'
12:55:47.356 [] [-zb-fs-workers-1] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-fs-workers': closed successfully
12:55:47.360 [] [-zb-actors-4] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-actors': closed successfully
12:55:47.472 [] [main] DEBUG io.camunda.zeebe.logstreams.snapshot - Closed database from '/tmp/junit5834930714341401105/runtime'.
```

</pre>
</details>

### PR DESCRIPTION
**Pull request by lenaschoenburg**
## Description
Taking a manual snapshot with `AsyncSnapshotDirector.forceSnapshot` used to be completely async.
Because this has made some tests flaky, we now return a future that is completed when the snapshot was taken (or skipped).
This allows us to wait for snapshots in tests.

I realize this is a bit of a maximalist solution to a flaky test issue so I'm also okay with closing this PR and just fixing it by increasing some timeouts.

## Related issues

closes #8686

### CODE REVIEW

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java [hunk 1]
```diff
@@ -56,7 +58,7 @@
   private Long lastWrittenPosition;
   private TransientSnapshot pendingSnapshot;
   private long lowerBoundSnapshotPosition;
-  private boolean takingSnapshot;
+  private CompletableActorFuture<Optional<PersistedSnapshot>> snapshotFuture;
```
Comment (Zelldon) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java:61]: :wrench: Tbh it looks a bit weird the combination of Future and Optional :sweat_smile:

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java [hunk 2]
```diff
@@ -114,37 +117,38 @@ private void createAsyncSnapshotDirectorOfReplayMode() {
...
-                assertThat(persistedSnapshotStore.getCurrentSnapshotIndex())
-                    .describedAs("New snapshot is taken")
-                    .isGreaterThan(oldSnapshot.getIndex()));
+    assertThat(secondSnapshot.join())
+        .describedAs("Second snapshot is taken")
+        .isPresent()
+        .get()
+        .extracting(PersistedSnapshot::getIndex, as(InstanceOfAssertFactories.LONG))
+        .describedAs("Second snapshot has a higher index")
+        .isGreaterThan(firstSnapshotIndex);
```
Comment (Zelldon) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java:150]: :wrench: Would be also nice to compare the snapshot with the snapshot for the snapshot store. Wdyt?

**Review by ChrisKujawa — COMMENTED** [review]
Thanks @oleschoenburg sorry that it take so long! In general I like it, but I'm a bit unsure with the optional maybe we can remove that. 

Additional I would like to see a test that we ensure the snapshot is equal to the snapshot which is part of the snapshot store.

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java (hunk 1 above)
Comment (oleschoenburg) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java:61]: I sort of agree but couldn't find a better solution. The optional is only needed for the case that we are already taking a snapshot and thus ignoring the request for a new one, and in the case where we haven't processed anything yet.

I think both cases don't really justify completing the future exceptionally.

Comment (oleschoenburg) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java:61]: I've tried out using sealed interfaces and records for the result type of the future but that made writing the assertions very cumbersome.

Comment (oleschoenburg) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java]: Okay, I've gone back to a future of a nullable snapshot which should be a bit more readable now :+1:

**Review by ChrisKujawa — APPROVED** [review]
Thanks @oleschoenburg :+1:

**Comment by Zelldon** [comment]
@oleschoenburg should we backport this? I suppose?

**Pull request merged**